### ISSUE TITLE
Research options for Passive DNS

### ISSUE THREAD

**Issue by markdboyd** [cloud-gov/product#2153]
From [EO M-21-31](https://www.whitehouse.gov/wp-content/uploads/2021/08/M-21-31-Improving-the-Federal-Governments-Investigative-and-Remediation-Capabilities-Related-to-Cybersecurity-Incidents.pdf):

> Federal agencies shall implement a Domain Name System (DNS) logging system that meets the requirements identified in Appendix C, including DNS requests made over encrypted DNS connections. Agencies shall implement accompanying analytics that allow for rapid identification of the host that sourced each DNS query. This capability shall be monitored and triaged. Federal agencies shall automate the production of a list of hostnames that are frequently accessed or looked up by legitimate users within their agency, but are not included in general top domain lists identified by CISA or available publicly or via subscription. Agencies should make that list automatically accessible to CISA or submit it to CISA daily via an acceptable automated mechanism

Passive DNS Log requirements from page 15 of the EO:

- Tuple (Rrname, Rrtype, Rdata) 
- Time_First 
- Time_Last 
- Count 
- Bailiwick 
- Sensor_Id 
- Zone_Time_First 
- Zone_Time_Last 
- Time_First_Ms 
- Time_Last_Ms 
- Origin 
- Count of Questions Asked by Source IP 
- Count of Questions Asked Overall 
- Count of Responses by Source IP 
- Query Size in Bytes 
- Response Size in Bytes 
- TTL per Record Returned 
- Request Was Made Via UDP, TCP or Both 
- Response Was Made Via UDP, TCP or Both 
- Passive DNS Source (Used to Identify Which Passive DNS Source Data Came From)

## Background

- [Presidential Memorandum M-21-31](https://www.whitehouse.gov/wp-content/uploads/2021/08/M-21-31-Improving-the-Federal-Governments-Investigative-and-Remediation-Capabilities-Related-to-Cybersecurity-Incidents.pdf).
- [Notes from discussion with assurance team about M-21-31](https://docs.google.com/document/d/11PcgXjs42hrYu8o91ep2vmdS5TkjmTw69Br0P1BvsBg/edit)
- [What is passive DNS?](https://www.spamhaus.com/resource-center/what-is-passive-dns-a-beginners-guide/)
  - "With special probes activated on the DNS resolver, after a cache miss has happened, it is possible to record the packets containing the answers to the client, along with the time & date stamp of when the query was made. This is Passive DNS data."
  - Essentially it is logging the responses of DNS queries

## Acceptance criteria

- [x] If feasible and not cost-prohibitive, a plan for enabling DNS logging of Route 53 domains is written. Plan should address how/if logs will be aggregated into main logging platform.
- [x] Implement way to identify top domains that are accessed from within the agency

## Notes

- We manage multiple domains via Route 53 in our commercial accounts (cloud.gov, *.app.cloud.gov, etc)

## Acceptance Criteria

- [x] Research options and cost for logging DNS queries from Route 53:
  - https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/query-logs.html
- [x] Research options and costs for logging DNS resolvers:
  - https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resolver-query-logs.html 
- If DNS logging appears reasonable and not cost-prohibitive, write plan and ticket to implement it

**Comment by dandersonsw** [comment]
Logging DNS queries from Route53:

Query logs can be sent to AWS CloudWatch, and will incur typical charges: https://aws.amazon.com/cloudwatch/pricing/

**Comment by dandersonsw** [comment]
Terraform doc for resolver query logs: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/route53_resolver_query_log_config

**Comment by markdboyd** [comment]
These are the values available in the resolver query logs:

https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resolver-query-logs-format.html

These values don't seem to match up exactly with what is expected by the EO

**Comment by markdboyd** [comment]
And here are the values available in route 53 query logs: https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/query-logs.html#query-logs-format

**Comment by markdboyd** [comment]
At this point, there does seem to be a clear path to turning on query logging in Route 53 for our domains, then exporting those logs to S3. From S3 we could then import the logs into our ELK platform.

However, one problem is that the Route 53 logs don't seem to contain all of the data that the EO requires. Specifically, I don't think that the Route 53 logs will provide these fields:

Sensor_Id
Zone_Time_First
Zone_Time_Last
Time_First_Ms
Time_Last_Ms
Query Size in Bytes
Response Size in Bytes
TTL per Record Returned

However, it's a bit hard to say because I don't know what these fields are and there are probably some terminology differences between the EO and Route 53.

Some of the aggregate fields, like "Count of Questions Asked by Source IP", I **think** we could get by aggregation queries once the data is in ELK.

I opened an AWS support case in commercial prod to see if they could provide any guidance.

But given the uncertainty about whether the Route 53 DNS logging will meet our needs, should we pursue enabling it anyway? I guess the adage that something is better than nothing might apply here, but we may not be able to meet the "letter of the law" for the EO.

@cloud-gov/platform-ops Thoughts?

**Comment by markdboyd** [comment]
@rcgottlieb found this document that seems to spell out the definitions of the fields that the EO is looking for: https://datatracker.ietf.org/doc/html/draft-dulaunoy-dnsop-passive-dns-cof#section-3

**Comment by markdboyd** [comment]
[The document that Robert shared](https://datatracker.ietf.org/doc/html/draft-dulaunoy-dnsop-passive-dns-cof#section-3) summarizes the following required fields for Passive DNS:

- Mandatory
  - rrname
  - rrtype
  - rdata
  - time_first
  - time_last
- Optional
  - count
  - bailiwick
- Additional
  - sensor_id
  - zone_time_first
  - zone_time_last
  - origin
  - time_first_ms
  - time_last_ms

This is how I would map the data from [Route53 query logs](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/query-logs.html#query-logs-format) to the EO requirements:

- Query timestamp - ?
- Hosted zone ID - roughly corresponds to bailiwick. could maybe use this plus the query_timestamp to generate Zone_Time_First and Zone_Time_Last?
- Query name - rrname
- Query type - rrtype
- Response code - nothing
- Layer 4 protocol - source protocol (Request Was Made Via UDP, TCP or Both)
- Route 53 edge location - roughly corresponds to sensor_id
- Resolver IP address - ?
- EDNS client subnet - origin

And his is how I would map the data from [Route53 resolver query logs](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resolver-query-logs-format.html) to the EO requirements:

- version - nothing
- account_id - nothing
- answers - rdata. length of array = count
- region - nothing
- vpc_id - nothing
- query_timestamp - could be useful for generating `time_first`, `time_last`, `time_first_ms`, `time_last_ms`
- query_name - rrname
- query_type - nothing
- query_class - nothing
- rcode - nothing
- answer_type - rrtype
- answer_class - nothing
- srcaddr - origin
- srcport - nothing
- transport - source protocol (Request Was Made Via UDP, TCP or Both)
- srcids - nothing
- instance - nothing
- resolver_endpoint - nothing?
- firewall_rule_group_id - nothing
- firewall_rule_action - nothing
- firewall_domain_list_id - nothing

Since the EO requirements for Passive DNS logging want a tuple (Rrname, Rrtype, Rdata) to identify each record, it seems like only the resolver query logs will provide the necessary data. 

However, resolver query logs [only capture DNS queries made from inside of a specific VPC](https://aws.amazon.com/blogs/aws/log-your-vpc-dns-queries-with-route-53-resolver-query-logs/), whereas I think we would want to capture DNS queries made by the public? Based on [this diagram](https://diagrams.fr.cloud.gov/10-1-network.html), I'm fairly sure that DNS queries for customer traffic would originate from outside of our VPCs, so I'm not sure how to effectively meet the EO requirements for public DNS queries.

@dandersonsw - Thoughts?

**Comment by markdboyd** [comment]
After talking to @pauldoomgov, I confirmed my understanding that:

- [AWS query logs](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/query-logs.htm) capture requests made from the public internet, but do not contain any information about how requests were resolved (i.e. no `rdata`)
-[AWS resolver query logs](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resolver-query-logs.html) will log the actual resolution of DNS queries, but **only those made from within the specified VPCs**

As for which of these services help us to meet EO M-21-31 requirements for Passive DNS, it's very hard to say. The EO seems to want full logging of DNS queries and answers for any domain that we manage in Route 53, which is not possible for DNS queries from the public internet. So at best we can only satisfy the EO for DNS queries originating from within our infrastructure.

From a security perspective, logging DNS queries that originate in our infrastructure is probably much more valuable. As @pauldoomgov pointed out during our conversation, these resolver DNS query logs could help identify outgoing traffic to malicious sites from our infrastructure and even the specific VMs where the queries originated.

So my thought would be that even though it doesn't fully satisfy the EO guidance on Passive DNS, we should enable DNS query resolver logging even if only for the security benefit alone.
